pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -507,104 +406,210 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
-			})
-			// update pr completed time
-			return nil
+			wrapErr := fmt.Errorf("Failed to validate pipelinerun %q with error %v", pr.Name, err)
+			pr.Status.MarkResourceFailed(ReasonFailedValidation, wrapErr)
+			return nil, nil, nil, controller.NewPermanentError(wrapErr)
+		}
+	}
+
+	return pipelineState, d, pipelineSpec, nil
+}
+
+// Push changes (if any) to the TaskRun status, labels and annotations to
+// TaskRun definition in ectd
+func (c *Reconciler) updateStatusLabelsAndAnnotations(pr, original *v1beta1.PipelineRun) error {
+
+	var updated bool
+	if !equality.Semantic.DeepEqual(original.Status, pr.Status) {
+		if _, err := c.updateStatus(pr); err != nil {
+			c.Logger.Warn("Failed to update PipelineRun status", zap.Error(err))
+			return fmt.Errorf("failed to update PipelineRun status: %s", err.Error())
 		}
+		updated = true
 	}
 
+	// When we update the status only, we use updateStatus to minimize the chances of
+	// racing any clients updating other parts of the Run, e.g. the spec or the labels.
+	// If we need to update the labels or annotations, we need to call Update with these
+	// changes explicitly.
+	if !reflect.DeepEqual(original.ObjectMeta.Labels, pr.ObjectMeta.Labels) || !reflect.DeepEqual(original.ObjectMeta.Annotations, pr.ObjectMeta.Annotations) {
+		if _, err := c.updateLabelsAndAnnotations(pr); err != nil {
+			c.Logger.Warn("Failed to update PipelineRun labels/annotations", zap.Error(err))
+			return fmt.Errorf("failed to update PipelineRun labels/annotations: %s", err.Error())
+		}
+		updated = true
+	}
+
+	if updated {
+		go func(metrics *Recorder) {
+			err := metrics.RunningPipelineRuns(c.pipelineRunLister)
+			if err != nil {
+				c.Logger.Warnf("Failed to log the metrics : %v", err)
+			}
+		}(c.metrics)
+	}
+
+	return nil
+}
+
+func (c *Reconciler) finishReconcileUpdateEmitEvents(pr, original *v1beta1.PipelineRun, beforeCondition *apis.Condition, previousError error) error {
+	afterCondition := pr.Status.GetCondition(apis.ConditionSucceeded)
+	events.Emit(c.Recorder, beforeCondition, afterCondition, pr)
+	err := c.updateStatusLabelsAndAnnotations(pr, original)
+	events.EmitError(c.Recorder, err, pr)
+	// If there was an error on update, we want to re-queue to re-try and update the status
+	// otherwise we return the original error, so we preserve the type (permanent or not)
+	if err != nil {
+		return multierror.Append(previousError, err).ErrorOrNil()
+	}
+	return previousError
+}
+
+func (c *Reconciler) updatePipelineResults(ctx context.Context, pr *v1beta1.PipelineRun) {
+
+	// TODO: Use factory func instead of hard-coding this once OCI images are supported.
+	resolver := &resources.LocalPipelineRefResolver{
+		Namespace:    pr.Namespace,
+		Tektonclient: c.PipelineClientSet,
+	}
+	pipelineMeta, pipelineSpec, err := resources.GetPipelineData(ctx, pr, resolver.GetPipeline)
+	if err != nil {
+		if ce, ok := err.(*v1beta1.CannotConvertError); ok {
+			pr.Status.MarkResourceNotConvertible(ce)
```

Comment:
im confused by why this function seems to repeat a lot of what `prepare` does - is it possible for `prepare` to be called before this function instead of duplicating the logic?